### PR TITLE
Pass X-Request-Id if present

### PR DESCRIPTION
**Pull request by paunik**
Pass the value from the X-Request-Id to the logger context

- pass the X-Request-Id header value if present, pass generated UUID if the header not present

Mentioned:
- https://github.com/vapor/vapor/issues/2508

Why:
 - We needed a way to trace router -> web -> worker on the logger on the heroku and we found this https://devcenter.heroku.com/articles/http-request-id

Inspiration:
- https://api.rubyonrails.org/classes/ActionDispatch/RequestId.html
- https://http.dev/x-request-id

### CODE REVIEW

File: Sources/Vapor/Request/Request.swift [hunk 1]
```diff
@@ -250,7 +250,16 @@ public final class Request: CustomStringConvertible {
...
         self.logger[metadataKey: "request-id"] = .string(id)
+=======
+        if let requestId = self.headers["X_REQUEST_ID"].first {
+            self.logger[metadataKey: "request-id"] = .string(requestId)
+        } else {
+            self.logger[metadataKey: "request-id"] = .string(UUID().uuidString)
+        }
+
+>>>>>>> Stashed changes
         self.byteBufferAllocator = byteBufferAllocator
```
Comment (gwynne) [inline Sources/Vapor/Request/Request.swift:263]: Please fixup this leftover merge conflict.

**Review by gwynne — CHANGES_REQUESTED** [review]
This looks fine, aside from the conflict markers left unresolved.

File: Sources/Vapor/Request/Request.swift (hunk 1 above)
Comment (paunik) [reply Sources/Vapor/Request/Request.swift:263]: on it

Comment (paunik) [reply Sources/Vapor/Request/Request.swift]: Should be done, tnx for the quick check, I should have seen this myself.

File: Tests/VaporTests/HTTPHeaderTests.swift [hunk 2]
```diff
@@ -201,6 +201,13 @@ final class HTTPHeaderTests: XCTestCase {
         )
     }
 
+    func testXRequestId() throws {
+        var headers = HTTPHeaders()
+        let xRequestId = UUID().uuidString
+        headers.replaceOrAdd(name: .xRequestId, value: xRequestId)
+        XCTAssertEqual(headers.first(name: "X-Request-Id"), xRequestId)
+    }
+
```
Comment (gwynne) [inline Tests/VaporTests/HTTPHeaderTests.swift:210]: Can you also add an end-to-end test which checks that the header is used when present in a request?

**Review by gwynne — APPROVED** [review]

**Review by dannflor — APPROVED** [review]
Beautamous

**Review by MahdiBM — COMMENTED** [review]
Honestly considering this header doesn't make an appearance in MDN, I wouldn't hard-code it like this 🤔

I would instead add a configuration option so other users can configure it "based on headers", and there they could specify the "X-Request-ID" header.

I know this complicates the solution, but we have "tracing" and "task-locals" for logging now which enable you to this yourself to some extent, and caring about this particular header seems out of scope for Vapor to me.

**Comment by gwynne** [comment]
@MahdiBM I disagree - `X-Request-ID` is not an official standard, true, but it is a de facto one. (Enough so, even for people to have gone to the trouble of noting that [RFC 6648](https://www.rfc-editor.org/rfc/rfc6648) deprecated the `X-` prefix on headers and remarked on the need to find a new name.) Unless you can cite links to any alternate names commonly used for the same purpose, configurability would just bloat the API surface and force users to do additional work without adding any real value.

**Pull request merged**